src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMapsFree/src/Client/ElectricityMapsFreeClient.cs

```diff
@@ -0,0 +1,195 @@
+using CarbonAware.DataSources.ElectricityMapsFree.Constants;
+using CarbonAware.DataSources.ElectricityMapsFree.Model;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using System.Diagnostics;
+using System.Net.Http.Headers;
+using System.Net.Mime;
+using System.Net;
+using System.Text.Json;
+using System.Web;
+
+namespace CarbonAware.DataSources.ElectricityMapsFree.Client;
+
+public class ElectricityMapsFreeClient : IElectricityMapsFreeClient
+{
+    private static readonly JsonSerializerOptions options = new JsonSerializerOptions(JsonSerializerDefaults.Web);
+
+    private static readonly HttpStatusCode[] RetriableStatusCodes = new HttpStatusCode[]
+    {
+        HttpStatusCode.Unauthorized,
+        HttpStatusCode.Forbidden
+    };
+
+    private readonly HttpClient client;
+
+    private IOptionsMonitor<ElectricityMapsFreeClientConfiguration> ConfigurationMonitor { get; }
+
+    private ElectricityMapsFreeClientConfiguration Configuration => this.ConfigurationMonitor.CurrentValue;
+
+    private static readonly ActivitySource Activity = new ActivitySource(nameof(ElectricityMapsFreeClient));
+
+    private ILogger<ElectricityMapsFreeClient> Log { get; }
+
+    public ElectricityMapsFreeClient(IHttpClientFactory factory, IOptionsMonitor<ElectricityMapsFreeClientConfiguration> configurationMonitor, ILogger<ElectricityMapsFreeClient> log)
+    {
+        this.client = factory.CreateClient(IElectricityMapsFreeClient.NamedClient);
+        this.ConfigurationMonitor = configurationMonitor;
+        this.Log = log;
+        Configuration.Validate();
+        this.client.BaseAddress = new Uri(this.Configuration.BaseUrl);
+        this.client.DefaultRequestHeaders.Accept.Clear();
+        this.client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue(MediaTypeNames.Application.Json));
+    }
+
+    /// <inheritdoc/>
+    public async Task<GridEmissionDataPoint> GetCurrentEmissionsAsync(string countryCodeAbbreviation)
+    {
+        Log.LogInformation("Requesting latest carbon intensity from zone {countryCode}", countryCodeAbbreviation);
+
+        var parameters = new Dictionary<string, string>()
+        {
+            { QueryStrings.countryCodeAbbreviation, countryCodeAbbreviation }
+        };
+
+        var tags = new Dictionary<string, string>()
+        {
+            { QueryStrings.countryCodeAbbreviation, countryCodeAbbreviation }
+        };
+
+        var result = await this.MakeRequestAsync(Paths.Latest, parameters, tags);
+        var emissionData = JsonSerializer.Deserialize<GridEmissionDataPoint>(result, options) ?? throw new ElectricityMapsFreeClientException($"Error getting latest carbon intensity for countryCode {countryCodeAbbreviation}");
+        if (emissionData == null || emissionData.Data.CarbonIntensity == null || emissionData.Data.Datetime == null)
+        {
+            Log.LogError("Region {countryCode} is not known", countryCodeAbbreviation);
```

Comment:
It's not clear to me why the 3 above or statements would lead to "region code is not known" error. Could you elaborate why?

If there's not a specific correlation, a generic error can be logged as well with the error message from the HTTP included

Reply:
`if (emissionData == null)` is just to avoid the warning that `emissionData` in the next 2 checks could be null. The other two instead are the real check, and they are null when you ask for a region name that the data source does not know. Like "aaaaaaaa". But if you really pass "aaaaaaaa", the error is easy to spot. The real point is helping the users figuring out what went wrong if they pass something like "US-CALI-CISO" (instead of the correct "US-CAL-CISO"). In that case the error is not obvious, and an error like "Region US-CALI-CISO is not known" should give you a good hint. Even more explicit would be adding "Is that a typo?", but I think it's excessive.
So the short answer is that it is an attempt to give a clearer error message.